ambry-router/src/main/java/com.github.ambry.router/EncryptDecryptThread.java

```diff
@@ -0,0 +1,264 @@
+        closePendingJobs(new GeneralSecurityException("Shutting down EncryptDecrypt Thread"));
+      }
+    } finally {
+      lock.unlock();
+    }
+    if (shutdownLatch != null) {
+      try {
+        shutdownLatch.await(10, TimeUnit.SECONDS);
+      } catch (InterruptedException e) {
+        logger.error("InterruptedException thrown while waiting for EncryptDecryptThread to shutdown");
+      }
+    }
+  }
+
+  /**
+   * Closes all pending jobs with the given {@code e}
+   * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callbacks for the pending jobs
+   */
+  private void closePendingJobs(GeneralSecurityException gse) {
+    JobInfo jobInfo = jobQueue.poll();
+    while (jobInfo != null) {
+      jobInfo.closeJob(gse);
+      jobInfo = jobQueue.poll();
+    }
+  }
+
+  /**
+   * JobInfo representing the job that needs processing
+   */
+  static abstract class JobInfo {
+    protected final BlobId blobId;
+    protected final ByteBuffer input;
+
+    JobInfo(BlobId blobId, ByteBuffer input) {
+      this.blobId = blobId;
+      this.input = input;
+    }
+
+    /**
+     * Processing of the job using {@code cryptoService} and {@code kms}.
+     * @param cryptoService the {@link CryptoService} to assist in encrypting and decrypting
+     * @param kms the {@link KeyManagementService} to fetch keys from
+     */
+    abstract void doOperation(CryptoService cryptoService, KeyManagementService kms);
+
+    /**
+     * Close the job with the given {@code gse}
+     * @param gse the {@link GeneralSecurityException} that needs to be set while invoking callback for the job
+     */
+    abstract void closeJob(GeneralSecurityException gse);
+  }
+
+  /**
+   * Callback to invoke for encrypt jobs
+   */
+  interface EncryptCallBack {
+    /**
+     * Callback to be invoked on completion of an encryption job.
+     * @param blobId {@link BlobId} for which encryption was requested
+     * @param encryptedKey {@link ByteBuffer} representing the encrypted per blob key
+     * @param encryptedContent {@link ByteBuffer} representing encrypted content
+     * @param gse {@link GeneralSecurityException} thrown while encrypting
+     */
+    void onCompletion(BlobId blobId, ByteBuffer encryptedKey, ByteBuffer encryptedContent,
+        GeneralSecurityException gse);
+  }
+
+  /**
+   * Callback to invoke for decrypt jobs
+   */
+  interface DecryptCallBack {
+    /**
+     * Callback to be invoked on completion of an decryption job.
+     * @param blobId {@link BlobId} for which encryption was requested
+     * @param decryptedContent {@link ByteBuffer} representing decrypted content
+     * @param gse {@link GeneralSecurityException} thrown while decrypting
+     */
+    void onCompletion(BlobId blobId, ByteBuffer decryptedContent, GeneralSecurityException gse);
+  }
+
+  /**
+   * Class representing an encrypt Job.
+   */
+  static class EncryptJobInfo extends JobInfo {
```

Comment:
I would take out the `Info` from these names. I think `EncryptJob`/`DecryptJob`/`Job` is good.